File: spring-angular/roma/v4/backend-ms1/src/main/java/ma/zs/stocky/service/impl/admin/stock/AchatAdminServiceImpl.java
package ma.zs.stocky.service.impl.admin.stock;


import ma.zs.stocky.bean.core.stock.Achat;
import ma.zs.stocky.dao.criteria.core.stock.AchatCriteria;
import ma.zs.stocky.dao.facade.core.stock.AchatDao;
import ma.zs.stocky.dao.specification.core.stock.AchatSpecification;
import ma.zs.stocky.service.facade.admin.stock.AchatAdminService;
import ma.zs.stocky.zynerator.service.AbstractServiceImpl;
import ma.zs.stocky.zynerator.util.ListUtil;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.ArrayList;





import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import ma.zs.stocky.service.facade.admin.stock.AchatItemAdminService ;
import ma.zs.stocky.bean.core.stock.AchatItem ;
import ma.zs.stocky.service.facade.admin.commun.ClientAdminService ;
import ma.zs.stocky.bean.core.commun.Client ;

import java.util.List;
@Service
public class AchatAdminServiceImpl extends AbstractServiceImpl<Achat, AchatCriteria, AchatDao> implements AchatAdminService {


    @Transactional(propagation = Propagation.REQUIRED, rollbackFor = Exception.class, readOnly = false)
    public Achat create(Achat t) {
        Achat saved= super.create(t);
        if (saved != null && t.getAchatItems() != null) {
                t.getAchatItems().forEach(element-> {
                    element.setAchat(saved);
                    achatItemService.create(element);
            });
        }
        return saved;

    }

    public Achat findWithAssociatedLists(Long id){
        Achat result = dao.findById(id).orElse(null);
        if(result!=null && result.getId() != null) {
            result.setAchatItems(achatItemService.findByAchatId(id));
        }
        return result;
    }
    @Transactional
    public void deleteAssociatedLists(Long id) {
        achatItemService.deleteByAchatId(id);
    }


    public void updateWithAssociatedLists(Achat achat){
    if(achat !=null && achat.getId() != null){
            List<List<AchatItem>> resultAchatItems= achatItemService.getToBeSavedAndToBeDeleted(achatItemService.findByAchatId(achat.getId()),achat.getAchatItems());
            achatItemService.delete(resultAchatItems.get(1));
            ListUtil.emptyIfNull(resultAchatItems.get(0)).forEach(e -> e.setAchat(achat));
            achatItemService.update(resultAchatItems.get(0),true);
    }
    }




    public Achat findByReferenceEntity(Achat t){
        return  dao.findByReference(t.getReference());
    }
    public void findOrSaveAssociatedObject(Achat t){
        if( t != null) {
            t.setClient(clientService.findOrSave(t.getClient()));
        }
    }

    public List<Achat> findByClientId(Long id){
        return dao.findByClientId(id);
    }
    public int deleteByClientId(Long id){
        return dao.deleteByClientId(id);
    }
    public long countByClientCin(String cin){
        return dao.countByClientCin(cin);
    }

    public List<Achat> findAllOptimized() {
        return dao.findAllOptimized();
    }





    public void configure() {
        super.configure(Achat.class, AchatSpecification.class);
    }


    @Autowired
    private AchatItemAdminService achatItemService ;
    @Autowired
    private ClientAdminService clientService ;

    public AchatAdminServiceImpl(AchatDao dao) {
        super(dao);
    }

}

File: nest/v3/backend-ms1/src/app/module/service/admin/stock/AchatAdminServiceImpl.ts
import {Injectable, NotFoundException} from "@nestjs/common";

import {AbstractServiceImpl} from "src/app/zynerator/service/AbstractServiceImpl";

import {AchatDao} from "src/app/shared/dao/facade/core/stock/AchatDao";
import {AchatCriteria} from "src/app/shared/dao/criteria/core/stock/AchatCriteria";
import {Achat} from "src/app/shared/bean/core/stock/Achat";
import {AchatDto} from "src/app/shared/dto/stock/AchatDto";

import {AchatItemAdminServiceImpl} from "src/app/module/service/admin/stock/AchatItemAdminServiceImpl";
import {AchatItem} from "src/app/shared/bean/core/stock/AchatItem";

@Injectable()
export class AchatAdminServiceImpl extends AbstractServiceImpl<Achat, AchatCriteria, AchatDao>{

    constructor(private readonly dao: AchatDao ,
                 private readonly achatItemService: AchatItemAdminServiceImpl ,
    ) {
        super(dao);
    }

    async save(item: Achat): Promise<Achat> {
        const saved = await this.dao.save(item);
        if (item.achatItems) {
            const savedAchatItems: AchatItem[] = [];
            for (const achatItem of item.achatItems) {
                achatItem.achat = saved;
                const savedAchatItem = await this.achatItemService.save(achatItem);
                savedAchatItems.push(savedAchatItem);
            }
            saved.achatItems = savedAchatItems;
        }
        return saved;
    }


    async update(item: Achat): Promise<Achat> {
        const saved = await this.dao.saveOrUpdate(item);
        return saved;
    }

    async updateMultiple(items: Achat[]): Promise<void> {
        if (items) {
            items.forEach(e => this.update(e))
        }
    }

    async  findAllOptimized(): Promise<AchatDto[]> {
        return this.dao.findAllOptimized();
    }

    async findAll(): Promise<Achat[]> {
        return this.dao.findAll();
    }

    async findById(id: number): Promise<Achat> {
        return this.dao.findById(id);
    }


    async deleteById(id: number): Promise<number> {
        const existing = await this.findById(id);
        if (!existing) {
            throw new NotFoundException(`Achat with ID ${id} not found.`);
        }
       await this.achatItemService.deleteByAchatId(id)
        const result = await this.dao.deleteById(existing.id);
        return result;
    }

    async deleteMultiple(items: Achat[]): Promise<Achat[]> {
        const deletedItems: Achat[] = [];
        for (const item of items) {
            await this.deleteById(item.id);
            deletedItems.push(item);
        }
        return deletedItems;
    }

    async findByClientId(id: number): Promise<Achat[]> {
        return this.dao.findByClientId(id);
    }

    async deleteByClientId(id: number): Promise<number> {
        return this.dao.deleteByClientId(id);
    }

    async findWithAssociatedLists(id: number): Promise<Achat> {
        const result = await this.dao.findById(id);
        if (result && result.id) {
          result.achatItems = await this.achatItemService.findByAchatId(result.id);
        }
    return result;
    }

    async updateWithAssociatedLists(item: Achat): Promise<Achat> {
            if (item && item.id) {
                //update  achatItems
                const oldAchatItems = await this.achatItemService.findByAchatId(item.id);
                const result = this.achatItemService.getToBeSavedAndToBeDeleted(oldAchatItems, item.achatItems);
                if (result && result.length === 2) {
                    await this.achatItemService.deleteMultiple(result[1]);
                    (result[0] || []).forEach((e) => e.achat = item);
                    await this.achatItemService.updateMultiple(result[0]);
                }

                return this.update(item);
        }
    }
}

